File: src/lib/fingerprint.js
/**
 * Huella digital de audiolibros: identidad estable y portable.
 *
 * Como cada dispositivo guarda su propia copia del archivo, no podemos usar el
 * id local (un UUID aleatorio) para saber que dos ficheros son el mismo libro.
 * La huella se calcula del contenido, así que el mismo archivo copiado al móvil
 * produce exactamente el mismo identificador.
 *
 * IMPORTANTE: este algoritmo es un contrato compartido con la app de Android.
 * Cualquier cambio aquí obliga a cambiar la implementación en Kotlin y deja
 * huérfano el progreso ya sincronizado. Ver docs/SYNC.md.
 *
 *   huella de pista = SHA-256( primer MiB || último MiB || tamaño en ASCII )
 *   huella de libro = SHA-256( huellas de sus pistas, ordenadas y unidas por \n )
 *
 * Solo se leen 2 MiB de cada archivo, así que un M4B de 1,7 GB se resuelve en
 * milisegundos en vez de tener que digerirlo entero.
 */

const CHUNK = 1024 * 1024 // 1 MiB

function toHex(buffer) {
  return Array.from(new Uint8Array(buffer))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('')
}

async function sha256Hex(bytes) {
  const digest = await crypto.subtle.digest('SHA-256', bytes)
  return toHex(digest)
}

/**
 * Huella de un único archivo de audio (Blob o File).
 * Devuelve 64 caracteres hexadecimales.
 */
export async function trackFingerprint(blob) {
  const size = blob.size
  const head = new Uint8Array(await blob.slice(0, Math.min(CHUNK, size)).arrayBuffer())
  // El último MiB solo se añade si no solapa con el primero; en archivos
  // pequeños el head ya cubre todo el contenido.
  const tailStart = Math.max(head.length, size - CHUNK)
  const tail =
    tailStart < size
      ? new Uint8Array(await blob.slice(tailStart, size).arrayBuffer())
      : new Uint8Array(0)

  const sizeBytes = new TextEncoder().encode(String(size))
  const payload = new Uint8Array(head.length + tail.length + sizeBytes.length)
  payload.set(head, 0)
  payload.set(tail, head.length)
  payload.set(sizeBytes, head.length + tail.length)

  return sha256Hex(payload)
}

/**
 * Huella de un libro completo a partir de las huellas de sus pistas.
 * Se ordenan alfabéticamente para que la identidad del libro no dependa del
 * criterio de ordenación de cada plataforma: si el móvil ordenase las pistas
 * de otra forma, la huella seguiría coincidiendo.
 */
export async function bookFingerprintFrom(trackFingerprints) {
  const joined = [...trackFingerprints].sort().join('\n')
  return sha256Hex(new TextEncoder().encode(joined))
}

/**
 * Calcula las huellas de todas las pistas de un libro y la del libro.
 * Devuelve { bookFingerprint, trackFingerprints } en el orden original.
 */
export async function fingerprintBook(blobs) {
  const trackFingerprints = []
  for (const blob of blobs) {
    trackFingerprints.push(await trackFingerprint(blob))
  }
  return {
    bookFingerprint: await bookFingerprintFrom(trackFingerprints),
    trackFingerprints,
  }
}

File: android/app/src/main/java/com/lumina/audiolibros/sync/Fingerprint.kt
package com.lumina.audiolibros.sync

import java.io.File
import java.io.RandomAccessFile
import java.security.MessageDigest

/**
 * Huella digital de audiolibros: identidad estable y portable.
 *
 * Contrato compartido con la app de escritorio. Cualquier cambio aquí debe
 * hacerse también en `src/lib/fingerprint.js` del proyecto raíz, o los dos
 * dispositivos dejarán de reconocer los mismos libros. La especificación
 * completa y los vectores de referencia están en `docs/SYNC.md`.
 *
 *   huella de pista = SHA-256( primer MiB || último MiB || tamaño en ASCII )
 *   huella de libro = SHA-256( huellas de sus pistas, ordenadas y unidas por \n )
 *
 * Solo se leen 2 MiB por archivo: un M4B de 1,76 GB se resuelve en ~15 ms.
 */
object Fingerprint {

    private const val CHUNK = 1024 * 1024

    /**
     * Origen de bytes con acceso aleatorio. Se abstrae porque en Android los
     * audios pueden venir de un `File` o de un `content://` del selector de
     * documentos, y el algoritmo debe ser el mismo en ambos casos.
     */
    interface ByteSource {
        val size: Long
        fun read(offset: Long, length: Int): ByteArray
    }

    class FileSource(private val file: File) : ByteSource {
        override val size: Long get() = file.length()
        override fun read(offset: Long, length: Int): ByteArray {
            val buffer = ByteArray(length)
            RandomAccessFile(file, "r").use { raf ->
                raf.seek(offset)
                raf.readFully(buffer)
            }
            return buffer
        }
    }

    fun sha256Hex(data: ByteArray): String =
        MessageDigest.getInstance("SHA-256")
            .digest(data)
            .joinToString("") { "%02x".format(it.toInt() and 0xFF) }

    /** Huella de una pista. Devuelve 64 caracteres hexadecimales en minúsculas. */
    fun ofTrack(source: ByteSource): String {
        val size = source.size
        val headLength = minOf(CHUNK.toLong(), size).toInt()
        val head = if (headLength > 0) source.read(0, headLength) else ByteArray(0)

        // El último MiB solo se añade si no solapa con el primero.
        val tailStart = maxOf(headLength.toLong(), size - CHUNK)
        val tail = if (tailStart < size) source.read(tailStart, (size - tailStart).toInt()) else ByteArray(0)

        val sizeBytes = size.toString().toByteArray(Charsets.US_ASCII)
        return sha256Hex(head + tail + sizeBytes)
    }

    fun ofTrack(file: File): String = ofTrack(FileSource(file))

    /** Huella de una pista ya cargada en memoria (usado sobre todo en tests). */
    fun ofBytes(content: ByteArray): String = ofTrack(object : ByteSource {
        override val size: Long get() = content.size.toLong()
        override fun read(offset: Long, length: Int): ByteArray =
            content.copyOfRange(offset.toInt(), offset.toInt() + length)
    })

    /**
     * Huella de un libro a partir de las de sus pistas. Se ordenan
     * alfabéticamente para que la identidad no dependa del criterio de
     * ordenación de cada plataforma.
     */
    fun ofBook(trackFingerprints: List<String>): String =
        sha256Hex(trackFingerprints.sorted().joinToString("\n").toByteArray(Charsets.UTF_8))
}
